client/allocrunner/taskrunner/template/template.go

```diff
@@ -494,6 +508,74 @@ func (tm *TaskTemplateManager) onTemplateRendered(handledRenders map[string]time
 		}
 	}
 
+	// process script execution concurrently
+	var wg sync.WaitGroup
+	for _, script := range scripts {
+		wg.Add(1)
+		go tm.processScript(script, &wg)
+	}
+	wg.Wait()
+}
+
+// failOnError is a helper function that produces a TaskKilling event and emits
+// a message
+func (tm *TaskTemplateManager) failOnError(script *structs.ChangeScript, msg string) {
+	if script.FailOnError {
+		tm.config.Lifecycle.Kill(context.Background(),
+			structs.NewTaskEvent(structs.TaskKilling).
+				SetFailsTask().
+				SetDisplayMessage(msg))
+	}
+}
+
+// processScript is used for executing change_mode script and handling errors
+func (tm *TaskTemplateManager) processScript(script *structs.ChangeScript, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	if tm.handle == nil {
+		failure_msg := fmt.Sprintf(
+			"Template failed to run script %v with arguments %v because task driver doesn't support the exec operation",
+			script.Command,
+			script.Args,
+		)
+		tm.config.Events.EmitEvent(structs.NewTaskEvent(structs.TaskHookFailed).
+			SetDisplayMessage(failure_msg))
+		tm.failOnError(script, failure_msg+" Task is being killed.")
```

Comment:
I think you could move the event emission to the error handling function as well, so it would look something like this:

```go
func (tm *TaskTemplateManager) handleScriptError(script *structs.ChangeScript, msg string) {
    ev := structs.NewTaskEvent(structs.TaskHookFailed).SetDisplayMessage(msg)
    tm.config.Events.EmitEvent(ev)

    if script.FailOnError {
        tm.config.Lifecycle.Kill(context.Background(), structs.NewTaskEvent(structs.TaskHookFailed).SetDisplayMessage("Template script failed, task is being killed"))
    }
}

func (tm *TaskTemplateManager) processScript(script *structs.ChangeScript, wg *sync.WaitGroup) {
    defer wg.Done()

    if handle == nil {
        tm.handleScriptError(fmt.Sprintf("..."))
        return
    }
    # ...
```

The `Task is being killed` message can then also be handled in the error handler. I don't think the kill message needs the full reason again.